R/rsa.R

```diff
@@ -1,3 +1,35 @@
+# createUniqueId creates a random hex string of length bytes.
+# Uses its own random state.
+#
+# createUniqueId is used to create the token (name) for a public/private
+# key-pair. We have used three approaches for token generation:
+#
+# 1. Random hex string using the incoming random state. Produced duplicate
+#    tokens when users forced a constant initial seed (set.seed(0)).
+#    https://github.com/rstudio/rsconnect/issues/221
+# 2. MD5 of the public key. Produced empty tokens when the MD5 algorithm
+#    is disabled (due to FIPS mode).
+#    https://github.com/rstudio/rsconnect/issues/378
+# 3. Random hex string with isolated random state.
+createUniqueId <- function(bytes) {
+  # Seed handling inspired by:
+  # http://www.cookbook-r.com/Numbers/Saving_the_state_of_the_random_number_generator/
+  if (exists(".Random.seed", .GlobalEnv)) {
+    oldseed <- .GlobalEnv$.Random.seed
+  } else {
+    oldseed <- NULL
+  }
+  on.exit(
+    if (!is.null(oldseed)) {
+      .GlobalEnv$.Random.seed <- oldseed
+    } else {
+      rm(".Random.seed", envir = .GlobalEnv)
+    }
+  )
```

Comment:
By convention, we always use `add = TRUE` with `on.exit` so that it's not possible to inadvertently remove existing exit handlers.

Reply:
Done. Thanks.